nestkernel/source_table.cpp

```diff
@@ -273,6 +273,62 @@ nest::SourceTable::resize_sources( const thread tid )
   sources_[ tid ].resize( kernel().model_manager.get_num_synapse_prototypes() );
 }
 
+bool
+nest::SourceTable::source_should_be_processed_( const thread rank_start, const thread rank_end, const Source& source )
+{
+  if ( source.is_processed() or source.is_disabled() )
+  {
+    // looks like we've processed this already, let's continue
+    return false;
+  }
+
+  const thread source_rank = kernel().mpi_manager.get_process_id_of_node_id( source.get_node_id() );
+
+  // determine whether this thread is responsible for this part of
+  // the MPI buffer; if not we just continue with the next iteration
+  // of the loop
+  if ( source_rank < rank_start or source_rank >= rank_end )
+  {
+    return false;
+  }
+
+  return true; // if we reach this we should process it
+}
+
+bool
+nest::SourceTable::next_entry_has_same_source( const SourceTablePosition& current_position,
+  const Source& current_source ) const
```

Comment:
I would suggest to replace the entire code by

```c++
const auto &local_sources = sources_[ current_position.tid ][ current_position.syn_id ];
const size_t next_lcid = current_position.lcid + 1;

return ( next_lcid < local_sources.size() 
  and local_sources[ next_lcid ].get_node_id() == current_source.get_node_id() );
```

Reply:
nice!